### DISCUSSION TITLE
How to think about migration of job triggers.

### DISCUSSION THREAD

**Discussion by drusellers** [quartznet/quartznet#1778] [Q&A]
I've been building application, so I'm at the beginning of the development learning curve for this product. I've been using Quartz.Net to build out the various recurring task and the library is  meeting every need that I have. I've started by initially scanning the database at application start and starting all of the jobs in-memory. This is working well because I only have one instance of my application running at a time. However, I'm interested in running two instance of the application to meet basic high-availability needs.

In my current in-memory based approach, I can easily change the triggers since they are rebuilt at the application start. When I switch to the database persisted model, I won't have to schedule them at start up anymore since they will already be scheduled in the database. Which is really nice.

However, while reading the [Best Practices](https://www.quartz-scheduler.net/documentation/best-practices.html) I don't see anything that talks about larger / more general "business practices" around managing schedules. Clearly I can build API routes to manage this, but that would be a decent amount of effort - and I wouldn't be surprised to hear that there is a fundamentally different approach to managing triggers over the life time of an application.

Concrete Example

I have a job that checks a file system every 15 min for new files to arrive. When files arrive it then processes them. Nothing too fancy. Today, each time the app boots up - that 15 min can easily be tweaked and changed. Including, as well, the Job and Trigger Keys as well.

Because I don't have anything that is tracking this relationship directly - I'm considering an Entity in my own business database that stores the job name, the key name, its trigger interval, and payload. Then if I change that entity, it is responsible for making the needed changes in the quartz system.

Thoughts?

**Comment by lahma** [comment]
Generally if you'd have some sort of admin interface to manage triggers, that could be a separate scheduler instance inside a web application or other user interface that is configured with `ZeroSizeThreadPool` and with scheduler instance that is never started. This will give you scheduler interface for maintenance - CRUD operations for triggers and jobs. Then you would have regular worker nodes that would just process schedules.

If you go with configuration file kind of approach (process restart may be permitted to refresh) you could take a look at [XMLSchedulingDataProcessor](https://github.com/quartznet/quartznet/blob/main/src/Quartz/Xml/XMLSchedulingDataProcessor.cs) which has logic to sync configuration changes. There's already [an issue about adding JSON version](https://github.com/quartznet/quartznet/issues/1755) . But generally the XML should already offer the constructs even though not as cool as JSON nowadays.

There's also [an example for XML scheduling](https://github.com/quartznet/quartznet/blob/main/src/Quartz.Examples/15_ConfigureJobSchedulingByUsingXmlConfigurations/ConfigureJobSchedulingByUsingXmlConfigurationsExample.cs) that would also allow polling the [XML file](https://github.com/quartznet/quartznet/blob/main/src/Quartz.Examples/quartz_jobs.xml) for changes and refreshes when changes found.

So there's some examples that you might want to examine, with own entities in database you would create some duplication of data maintenance but of course also that would decouple from Quartz better. Maybe XML scheduling processor  shows some ideas how to sync such solution.

**Reply by drusellers** [reply]
The pattern of a `ZeroSizeThreadPool` would be a good pattern for the docs site. 👍 

So, I guess I've done a bad job of getting at the crux of my question.

Let's say I have 1,000s of jobs. And I've decided to re-model the JobKey Name and Group (or TriggerKey) to better represent something.

Do you have any patterns / suggestions for doing that rename? How do I take a given job instance and rename its key? I'm thinking that I deploy the change with the new structure and then have a job that runs and takes a job matching the old naming style, unschedules it, then reschedules it with the new naming. 🤔 

I'm currently using the JobKey.Name + Group to provide reporting value and maybe that's not the best way to go.

**Reply by lahma** [reply]
Ah, now I understand, no worries. Yes this kind of batch rename might be a bit hard to achieve *transactionally*. General advice is not to touch Quartz DB tables but this is something that might warrant for it and might be easier to script and deploy as migration. It won't be trivial though as you would need to drop unique constraints and bring them back after update.

The underlying problem really is that Quartz.NET has the name and group as primary keys - I wish I had changed that when porting but a bit too late - unless a painful migration step will be added for next version. 

Job and trigger groups are usually are more powerful feature as they allow mass-modifications like pausing and resuming. 

If it's just reporting value, maybe you could utilize some helper tables linking to names and groups to provide the reporting data. This would allow re-grouping or categorizing already scheduled data.

**Reply by drusellers** [reply]
Ok, that tracks with my own thinking at least. 

Ahhhh, I hadn't thought of using job and trigger groups in that regards. Mass pausing and starting. That's really interesting and I may need to rethink my approach there yet again. Would be nice to stop an entire customer's jobs with a simple call if they stop paying.

**Reply by lahma** [reply]
Yes,  especially as there are group and name matchers in API which perform best as StartsWith due to SQL indexes. It might be a good idea to design granularity so that specifity is from left to right. Happy designing! 🙂
